src/proof/alethe/alethe_printer.cpp

```diff
@@ -0,0 +1,265 @@
+ * Top contributors (to current version):
+ *   Hanna Lachnitt
+ *
+ * This file is part of the cvc5 project.
+ *
+ * Copyright (c) 2009-2021 by the authors listed in the file AUTHORS
+ * in the top-level source directory and their institutional affiliations.
+ * All rights reserved.  See the file COPYING in the top-level source
+ * directory for licensing information.
+ * ****************************************************************************
+ *
+ * The module for printing Alethe proof nodes.
+ */
+
+#include "proof/alethe/alethe_printer.h"
+
+#include <iostream>
+#include <unordered_map>
+
+#include "proof/alethe/alethe_proof_rule.h"
+
+namespace cvc5 {
+
+namespace proof {
+
+AletheProofPrinter::AletheProofPrinter() {}
+
+void AletheProofPrinter::print(std::ostream& out,
+                               std::shared_ptr<ProofNode> pfn)
+{
+  Trace("alethe-printer") << "- Print proof in Alethe format. " << std::endl;
+  std::unordered_map<Node, std::string> assumptions;
+  const std::vector<Node>& args = pfn->getArguments();
+  // Special handling for the first scope
+  // Print assumptions and add them to the list but do not print anchor.
+  for (size_t i = 3, size = args.size(); i < size; i++)
+  {
+    Trace("alethe-printer") << "... print assumption " << args[i] << std::endl;
+    out << "(assume a" << i - 3 << " " << args[i] << ")\n";
+    assumptions[args[i]] = "a" + std::to_string(i - 3);
+  }
+
+  // Then, print the rest of the proof node
+  int start_t = 1;
+  printInternal(out, pfn->getChildren()[0], assumptions, {}, "", start_t);
+}
+
+std::string AletheProofPrinter::printInternal(
+    std::ostream& out,
+    std::shared_ptr<ProofNode> pfn,
+    std::unordered_map<Node, std::string> assumptions,
+    std::unordered_map<Node, std::string> steps,
+    std::string current_prefix,
+    int& current_step_id)
+{
+  int step_id = current_step_id;
+  std::vector<std::string> current_assumptions;
+  const std::vector<Node>& args = pfn->getArguments();
+
+  // If the proof node is untranslated a problem might have occured during
+  // postprocessing
+  if (args.size() < 3 || pfn->getRule() != PfRule::ALETHE_RULE)
+  {
+    Trace("alethe-printer")
+        << "... printing failed! Encountered untranslated Node. "
+        << pfn->getResult() << " " << pfn->getRule() << " "
+        << " / " << args << std::endl;
+    return "";
+  }
+
+  // Get the alethe proof rule
+  AletheRule arule = getAletheRule(args[0]);
+
+  // In case the rule is an anchor it is printed before its children.
+  if (arule == AletheRule::ANCHOR_SUBPROOF || arule == AletheRule::ANCHOR_BIND)
+  {
+    // Look up if subproof has already been printed
+    auto it = steps.find(args[2]);
+    if (it != steps.end())
+    {
+      Trace("alethe-printer")
+          << "... subproof is already printed " << pfn->getResult() << " "
+          << arule << " / " << args << std::endl;
+      return it->second;
+    }
```

Comment:
you're checking this but I'm not sure where you are adding to the map?

Reply:
That's because I did not add it :sweat_smile: It is fixed now.